libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1459,24 +1459,25 @@ bool TypeChecker::visit(Conditional const& _conditional)
 	_conditional.trueExpression().accept(*this);
 	_conditional.falseExpression().accept(*this);
 
-	Type const* trueType = type(_conditional.trueExpression())->mobileType();
-	Type const* falseType = type(_conditional.falseExpression())->mobileType();
-
+	Type const* trueType =  type(_conditional.trueExpression());
+	Type const* falseType = type(_conditional.falseExpression());
+	Type const* trueMobileType = trueType->mobileType();
+	Type const* falseMobileType = falseType->mobileType();
 	Type const* commonType = nullptr;
 
-	if (!trueType)
-		m_errorReporter.typeError(9717_error, _conditional.trueExpression().location(), "Invalid mobile type in true expression.");
-	else
+	if (trueMobileType)
 		commonType = trueType;
-
-	if (!falseType)
-		m_errorReporter.typeError(3703_error, _conditional.falseExpression().location(), "Invalid mobile type in false expression.");
 	else
+		m_errorReporter.typeError(9717_error, _conditional.trueExpression().location(), "Invalid mobile type in true expression.");
+
+	if (falseMobileType)
 		commonType = falseType;
+	else
+		m_errorReporter.typeError(3703_error, _conditional.falseExpression().location(), "Invalid mobile type in false expression.");
 
-	if (!trueType && !falseType)
+	if (!trueMobileType && !falseMobileType)
 		BOOST_THROW_EXCEPTION(FatalError());
-	else if (trueType && falseType)
+	else if (trueMobileType && falseMobileType)
 	{
 		commonType = Type::commonType(trueType, falseType);
```

Comment:
So this is the main "change" here, right? So if we have types that are not their own mobile types, the common type will be computed from the non-mobile types instead of the mobile types, right?
What are the common types in these cases?

Reply:
To be honest, I still cannot really grasp the consequences of this change.

Reply:
Yes, this is the main change. The common type does not change:
- It is a mobile type of `a` if  `b` is implicitly convertible to mobile type of `a`. 
- It is a mobile type of `b` if `a` is implicitly convertible to mobile type of `b`. 
- Otherwise it is `nullptr`. 

Please notice that `Type::commonType` function checks if conversion between type of one parameter and mobile type of another parameter is possible. It converts to mobile type when it is necessary and also returns mobile type. However, when we convert parameters to their mobile type before we pass them to the function, we loose information that might be crucial. For instance, we loose information about a literal type exact value. This is why I think we should call `Type::commonType` with full types.